### PR TITLE
Add test cases for cobalt::loader::FetcherCache

### PR DESCRIPTION
**Pull request by borongc**
Test FetcherCache with corner cases, and try to catch crash issue.

b/270993319

### CODE REVIEW

File: cobalt/loader/fetcher_cache_test.cc [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+  stub_fetcher_factory2.fetcher->PostFetcherCallback(headers, data.c_str(),
+                                                     data.size());
+
+  // Let the message loop run for 200ms to allow enough time for the async
+  // script to fire the callback.
+  base::ThreadTaskRunnerHandle::Get()->PostDelayedTask(
+      FROM_HERE, base::BindOnce(run_loop.QuitClosure()),
+      base::TimeDelta::FromMilliseconds(200));
+  run_loop.Run();
+  EXPECT_EQ(data.capacity() * 2, fetcher_factory->size());
```
[the remark this answers is not in the document]
Comment (borongc) [reply cobalt/loader/fetcher_cache_test.cc]: 1. Add `MaxFetcherCache` test case.
2. Add `DisorderedSameFetcherCache` and `DisorderedDiffFetcherCache`.

File: cobalt/loader/fetcher_cache.cc [hunk 2]
```diff
@@ -348,6 +348,9 @@ void FetcherCache::OnFetchSuccess(
 
   total_size_ += entry->capacity();
   ++count_resources_cached_;
+  // TODO(b/270993319): For debugging cache integrity issues in production
+  //                    only, remove after identifying the root cause.
+  CHECK_EQ(count_resources_cached_, (int)cache_entries_.size());
```
Comment (borongc) [inline cobalt/loader/fetcher_cache.cc:353]: Add `CHECK_EQ` for potential inconsistent cache_entries_.

Comment (borongc) [reply cobalt/loader/fetcher_cache.cc]: Move to PR https://github.com/youtube/cobalt/pull/1983.

File: cobalt/loader/fetcher_cache_test.cc [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+
+TEST_F(FetcherCacheTest, MakeFetcher) {
+  OngoingFetcherFactory fetcher_factory("cache_name", 1024 * 1024);
+  auto fetcher_creator = fetcher_factory.MakeCachedFetcherCreator(
+      GURL("https://example.com"), base::Bind(&CreateFecter));
+  EXPECT_EQ(0, fetcher_factory.size());
+}
+
+TEST_F(FetcherCacheTest, SunnyDay) {
+  base::test::ScopedTaskEnvironment scoped_task_environment_;
```
Comment (sideb0ard) [inline cobalt/loader/fetcher_cache_test.cc:142]: You can move the scoped_task_environment_ to be a member of the test fixture, and save adding it into each individual test.

File: cobalt/loader/fetcher_cache_test.cc [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+  StubFetcherHandler handler;
+  StubFetcherFactory stub_fetcher_factory;
+  auto cb = stub_fetcher_factory.GetFetcherCreator();
+
+  OngoingFetcherFactory fetcher_factory("cache_name", 1024 * 1024);
```
Comment (sideb0ard) [inline cobalt/loader/fetcher_cache_test.cc:147]: this fetcher_factory could also be a member of the test fixture - might need to use curly brace initialization:
OngoingFetcherFactory fetcher_factory{"cache_name", 1024 * 1024};

File: cobalt/loader/fetcher_cache_test.cc [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+
+   private:
+    scoped_refptr<FetcherCache> fetcher_cache_;
+  };
+
+  const std::string data_ = "Test string! Test string! Test!!";
+};
+
+TEST_F(FetcherCacheTest, CtorAndDtor) {
+  OngoingFetcherFactory fetcher_factory("cache_name", 1024 * 1024);
```
Comment (sideb0ard) [inline cobalt/loader/fetcher_cache_test.cc:130]: the "1024 * 1024" is used throughout - you could make this a static constexpr

**Review by sideb0ard — APPROVED** [review]
Looks great! a few minor comments.

File: cobalt/loader/fetcher_cache_test.cc [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+    fetcher = new StubFetcher(handler);
+    return base::WrapUnique<Fetcher>(fetcher);
+  }
+
+  Loader::FetcherCreator GetFetcherCreator() {
+    return base::Bind(&StubFetcherFactory::Create, base::Unretained(this));
+  }
+};
+
+std::unique_ptr<Fetcher> CreateFecter(Fetcher::Handler* handler) {
```
[the remark this answers is not in the document]
Comment (borongc) [reply cobalt/loader/fetcher_cache_test.cc:93]: Fixed.

File: cobalt/loader/fetcher_cache_test.cc (hunk 3 above)
Comment (borongc) [reply cobalt/loader/fetcher_cache_test.cc:142]: Done.

File: cobalt/loader/fetcher_cache_test.cc (hunk 5 above)
Comment (borongc) [reply cobalt/loader/fetcher_cache_test.cc:130]: Done.

File: cobalt/loader/fetcher_cache_test.cc (hunk 4 above)
Comment (borongc) [reply cobalt/loader/fetcher_cache_test.cc]: Done.
